homeassistant/components/voip/__init__.py

```diff
@@ -0,0 +1,65 @@
+"""The Voice over IP integration."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Callable
+import logging
+
+from voip_utils import SIP_PORT
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_IP_ADDRESS
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+from .voip import HassVoipDatagramProtocol
+
+_LOGGER = logging.getLogger(__name__)
+_IP_WILDCARD = "0.0.0.0"
+
+__all__ = [
+    "DOMAIN",
+    "async_setup_entry",
+    "async_unload_entry",
+]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up VoIP integration from a config entry."""
+    ip_address = entry.data[CONF_IP_ADDRESS]
+    _LOGGER.debug(
+        "Listening for VoIP calls from %s (port=%s)",
+        ip_address,
+        SIP_PORT,
+    )
+    hass.data[DOMAIN] = await _create_sip_server(
+        hass,
+        lambda: HassVoipDatagramProtocol(hass, {str(ip_address)}),
+    )
+
+    return True
+
+
+async def _create_sip_server(
+    hass: HomeAssistant,
+    protocol_factory: Callable[
+        [],
+        asyncio.DatagramProtocol,
+    ],
+) -> asyncio.DatagramTransport:
+    transport, _protocol = await hass.loop.create_datagram_endpoint(
```

Comment:
This looks like something that should be done in a 3rd party library.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices

Abstract the asyncio datagram protocol interface so the library user doesn't need to know about that detail.